core/src/main/scala/org/apache/spark/internal/io/HadoopMapReduceCommitProtocol.scala

```diff
@@ -186,7 +186,9 @@ class HadoopMapReduceCommitProtocol(
         logDebug(s"Clean up default partition directories for overwriting: $partitionPaths")
         for (part <- partitionPaths) {
           val finalPartPath = new Path(path, part)
-          fs.delete(finalPartPath, true)
+          if (!fs.delete(finalPartPath, true) && !fs.exists(finalPartPath.getParent)) {
```

Comment:
why we only create the parent dir if we fail to delete the `finalPartPath`?

Reply:
@cloud-fan this is to follow the behavior of HDFS rename spec: it requires the parent to be present. If we create finalPartPath directly, then it will cause another wired behavior in rename when the dst path already exists. From the HDFS spec I shared above: " If the destination exists and is a directory, the final destination of the rename becomes the destination + the filename of the source path".  We have confirmed this in our production cluster, and resulted in the current solution to only create parent dir which follows the HDFS spec exactly,

Reply:
isn't it more clear to just do `assert(fs.delete(finalPartPath, true))`?

Reply:
sorry I don't get it, if the delete success, your patch doesn't change anything, do you mean HDFS will always fail to delete `finalPartPath`?

Reply:
I think the problem here is we didn't check whether the `finalPartPath` exists, and we shall actually check that before rename.

Reply:
I feel the code here is not safe. We may fail to delete if `finalPartPath` doesn't exist, or there are some real failures. We should make sure `finalPartPath` doesn't exist before renaming.

Reply:
BTW we should add comments around here to explain all these stuff.

Reply:
+1 on add comments.

Reply:
The FileSystem API spec on delete says "Code SHOULD just call delete(path, recursive) and assume the destination is no longer present". Referring to its detailed spec, the only case that we may get false from delete would be <code>finalPartPath</code> does not exist. Other failures should result in exception.  When <code>finalPartPath</code> does not exist, which is an expected case, we should only check if the parent of <code>finalPartPath</code> does not exist because otherwise we will have problem in rename according to its spec.  Please advise if you guys think we still should double-check <code>finalPartPath</code> before rename

Reply:
ah makes sense, let's add some comment to summary these discussions.